### ISSUE TITLE
All settings and playlists lost after application crash

### ISSUE THREAD

**Issue by dalto8** [fooyin/fooyin#127]
fooyin crashed with this:
```
Jun 01 12:58:49 hostname systemd-coredump[335609]: Process 65715 (fooyin) of user 1000 dumped core.
                                                #7  0x000075ee767e061c _ZN6Fooyin13PlaylistModel13updateHeadersEv (libfooyin_gui.so.0.0.0 + 0x11261c)
                                                #8  0x000075ee767e759e _ZN9QtPrivate11FunctorCallINS_11IndexesListIJLi0EEEENS_4ListIJN6Fooyin11PendingDataEEEEvZNS4_13PlaylistModelC4EPNS4_12MusicLibraryEPNS4_16PlayerControllerEPNS4_15SettingsManagerEP7QObjectEUlS5_E0_E4callERSG_PPv.lto_priv.0 (libfooyin_gui.so.0.0.0 + 0x11959e)
                                                #21 0x000075ee76781825 _ZN6Fooyin22PropertiesDialogWidget5applyEv (libfooyin_gui.so.0.0.0 + 0xb3825)
                                                #22 0x000075ee76781f0d _ZN6Fooyin22PropertiesDialogWidget6acceptEv (libfooyin_gui.so.0.0.0 + 0xb3f0d)
                                                #47 0x000061930dbf3fe7 main (fooyin + 0x3fe7)
                                                #50 0x000061930dbf4495 _start (fooyin + 0x4495)
Jun 01 12:58:49 hostname systemd[2396]: app-org.fooyin.fooyin-89e0ac13287f44c18b476062c8c6f039.scope: Consumed 32min 31.327s CPU time.
Jun 01 12:58:50 hostname drkonqi-coredump-processor[335610]: "/usr/bin/fooyin" 65715 "/var/lib/systemd/coredump/core.fooyin.1000.0113cb87111747ac9651af805f070c05.65715.1717264721000000.zst"
Jun 01 12:58:50 hostname drkonqi-coredump-launcher[335650]: Unable to find file for pid 65715 expected at "kcrash-metadata/fooyin.0113cb87111747ac9651af805f070c05.65715.ini"
```

This would not be a huge deal except that when I reopened it all my settings and playlist contents were completely lost.  The only thing that remained was the names of my playlists.

Given how configurable the application this is....not ideal.  Even if the settings were lost I would have expected the playlist contents to be more durable as rebuilding all the playlists is a major effort.

As a side note, where is the database stored?

**Comment by ludouzi** [comment]
Thanks for the core dump. I'll look into it. Did it crash after editing tags in the tag editor by any chance?

The DB is normally found in `.local/share/fooyin`. When creating a playlist, it is saved immediately. Tracks (and the current layout and settings) are currently only saved on a clean exit. I'll see about improving upon that.

**Comment by dalto8** [comment]
> Did it crash after editing tags in the tag editor by any chance?

Yes

> Tracks (and the current layout and settings) are currently only saved on a clean exit.

That explains it.  I would say that tracks, at the very least, should be saved more often.

Luckily, I had exported my layout a few minutes before the crash so I had a copy of it.  Because that took me forever to get the way I wanted it.

Thanks for all your work on this.  It is coming together really nicely.  I am thoroughly impressed with it.

**Comment by ludouzi** [comment]
I managed to reproduce and fix a few crashes when updating tracks in a playlist with https://github.com/ludouzi/fooyin/commit/72e2a57350bdbe2b37f0aef7af32f7e3cfc5c9ef. The underlying playlist model is very complex however, so I'll keep testing to make sure I've covered all related cases.

> That explains it. I would say that tracks, at the very least, should be saved more often.

Agreed. I think the best way to handle that would be to start/restart a timer whenever tracks are changed in a playlist. Once it times out, save changed playlists to the DB.

> Thanks for all your work on this. It is coming together really nicely. I am thoroughly impressed with it.

Thanks. it's great to hear such feedback.
